spec/unit/govuk_index/publishing_event_worker_spec.rb

```diff
@@ -1,166 +1,275 @@
+      expect(Services.statsd_client).to receive(:increment).with('govuk_index.elasticsearch.delete')
+      subject.perform([['routing.key', payload1], ['routing.key', payload_delete]])
+    end
+
+    context 'when all messages fail' do
+      before do
+        allow(actions).to receive(:save).twice
+        allow(actions).to receive(:commit).and_return(
+          'items' => [{ 'index' => { 'status' => 500 } }, { 'index' => { 'status' => 500 } }]
+        )
+        allow(Services.statsd_client).to receive(:increment)
+      end
+
+      it 'will reprocess the entire batch using ES retry mechanism' do
+        expect(Services.statsd_client).to receive(:increment).with('govuk_index.sidekiq-consumed').twice
+
+        expect {
+          subject.perform([['routing.key', payload1], ['routing.key', payload2]])
+        }.to raise_error(GovukIndex::ElasticsearchRetryError)
+      end
+
+      it 'will notify for each message that fails' do
+        expect(Services.statsd_client).to receive(:increment).with('govuk_index.elasticsearch.index_error').twice
+        expect(Services.statsd_client).to receive(:increment).with('govuk_index.sidekiq-retry')
+
+        expect {
+          subject.perform([['routing.key', payload1], ['routing.key', payload2]])
+        }.to raise_error(GovukIndex::ElasticsearchRetryError)
+      end
+
+      it 'will notify that the batch failed' do
+        expect(Services.statsd_client).to receive(:increment).with('govuk_index.sidekiq-retry')
+
+        expect {
+          subject.perform([['routing.key', payload1], ['routing.key', payload2]])
+        }.to raise_error(GovukIndex::ElasticsearchRetryError)
+      end
+    end
+
+    context 'elasticsearch fails during processing for some messages' do
+      before do
+        allow(actions).to receive(:save).twice
+        allow(actions).to receive(:commit).and_return(
+          'items' => [{ 'index' => { 'status' => 200 } }, { 'index' => { 'status' => 500 } }]
+        )
+        allow(Services.statsd_client).to receive(:increment)
+        # allow(GovukIndex::PublishingEventWorker).to receive(:perform_async)
+      end
+
+      it 'will raise an error so that sidekiq retries the entire batch' do
+        expect {
+          subject.perform([['routing.key', payload1], ['routing.key', payload2]])
+        }.to raise_error(GovukIndex::ElasticsearchRetryError)
+      end
+
+      it 'will notify for each message that fails' do
+        expect(Services.statsd_client).to receive(:increment).with('govuk_index.elasticsearch.index_error')
+        expect(Services.statsd_client).to receive(:increment).with('govuk_index.elasticsearch.index')
+        expect {
+          subject.perform([['routing.key', payload1], ['routing.key', payload2]])
+        }.to raise_error(GovukIndex::ElasticsearchRetryError)
+      end
+    end
+
+    it "raises an error is the number of response item does not match the number actions performed" do
```

Comment:
Not quite sure I follow the test name. Is it supposed to be "raises an error if the number of response items does not match the number of actions performed"?

Reply:
maybe "..actions requested" is a better name.

Basically when I send 5 actions in the bulk message to ES if I get 4 or 6 responses then something has gone wrong.

Reply:
updated as discussed